src/main/java/gregtech/api/fluids/definition/MaterialFluidDefinition.java

```diff
@@ -0,0 +1,257 @@
+     *
+     * @throws UnsupportedOperationException This is unsupported.
+     */
+    @Nonnull
+    @Override
+    public Fluid constructFluid(@Nonnull String fluidName) throws UnsupportedOperationException {
+        throw new UnsupportedOperationException("Use constructFluid(material) instead.");
+    }
+
+    @Override
+    public void setTemperature(int temperature) {
+        super.setTemperature(temperature);
+        this.needsTemperatureInference = temperature == -1;
+    }
+
+    /**
+     * @param material the material to get the registry name for
+     * @return the registry name associated with this definition and the material
+     */
+    @Nonnull
+    public String getRegistryName(@Nonnull Material material) {
+        return registryNameOverride != null ? registryNameOverride : getType().getFluidNameForMaterial(material);
+    }
+
+    /**
+     * @return the fluid type for this fluid
+     */
+    @Nonnull
+    public FluidType getType() {
+        return type;
+    }
+
+    /**
+     * @param type the type to set
+     */
+    public void setType(@Nonnull FluidType type) {
+        this.type = type;
+    }
+
+    /**
+     * @return the material icon type for the still texture
+     */
+    @Nonnull
+    public MaterialIconType getStillIconType() {
+        return stillIconType;
+    }
+
+    /**
+     * @param stillIconType the icon type to set
+     */
+    public void setStillIconType(@Nonnull MaterialIconType stillIconType) {
+        this.stillIconType = stillIconType;
+    }
+
+    /**
+     * @return the material icon type for the flowing texture
+     */
+    @Nonnull
+    public MaterialIconType getFlowingIconType() {
+        return flowingIconType;
+    }
+
+    /**
+     * @param flowingIconType the icon type to set
+     */
+    public void setFlowingIconType(@Nonnull MaterialIconType flowingIconType) {
+        this.flowingIconType = flowingIconType;
+    }
+
+    public static class Builder extends FluidDefinition.AbstractBuilder<MaterialFluidDefinition, Builder> {
+
+        protected FluidType type;
+        protected MaterialIconType stillIconType;
+        protected MaterialIconType flowingIconType;
+        protected boolean hasCustomTexture = false;
+        protected String registryNameOverride = null;
+
+        /**
+         * @param type   the fluid type for this fluid
+         * @param state the state of the fluid
+         */
+        public Builder(@Nonnull FluidType type, @Nonnull FluidState state) {
+            super(state);
+            this.type = type;
+            this.stillIconType = state.getStillIconType();
+            this.flowingIconType = state.getStillIconType();
+        }
+
+        /**
+         * Set the still icon type for this fluid
+         *
+         * @param stillIconType the icon type to set
+         */
+        @Nonnull
+        public Builder still(@Nonnull MaterialIconType stillIconType) {
```

Comment:
These ones use MaterialIconTypes instead of raw ResourceLocations. These are a lot more useful when adding fluids for materials, particularly in the case that it will grab the correct one for the material's iconset.